apiserver/facades/client/spaces/spaces_rename.go

```diff
@@ -0,0 +1,255 @@
+/*
+ * // Copyright 2020 Canonical Ltd.
+ * // Licensed under the AGPLv3, see LICENCE file for details.
+ */
+
+package spaces
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/common/networkingcommon"
+	"gopkg.in/juju/names.v3"
+	"gopkg.in/mgo.v2/txn"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	jujucontroller "github.com/juju/juju/controller"
+	"github.com/juju/juju/core/constraints"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/core/settings"
+	"github.com/juju/juju/state"
+)
+
+// RenameSpaceModelOp describes a model operation for committing a branch.
+type RenameSpaceModelOp interface {
+	state.ModelOperation
+}
+
+// RenameSpaceCompleteOps describes a space that can be renamed..
+type RenameSpace interface {
+	Refresh() error
+	Id() string
+	Name() string
+	RenameSpaceCompleteOps(toName string) ([]txn.Op, error)
+}
+
+// CommitBranchState describes state operations required
+// to execute the renameSpace operation.
+// * This allows us to indirect state at the operation level instead of the
+// * whole API level as currently done in interface.go
+type RenameSpaceState interface {
+	// ControllerConfig returns current ControllerConfig.
+	ControllerConfig() (jujucontroller.Config, error)
+
+	//ConstraintsBySpace returns current constraints using the given spaceName.
+	ConstraintsBySpace(spaceName string) (map[string]constraints.Value, error)
+
+	// SpaceByName returns the Juju network space given by name.
+	SpaceByName(name string) (networkingcommon.BackingSpace, error)
+
+	// ControllerSettingsGlobalKey() returns the global controller settings key.
+	ControllerSettingsGlobalKey() string
+
+	// GetConstraintsOps gets the database transaction operations for the given constraints.
+	// Cons is  a map keyed by the DocID.
+	GetConstraintsOps(cons map[string]constraints.Value) ([]txn.Op, error)
+}
+
+// Settings describes methods for interacting with settings to apply
+// branch-based configuration deltas.
+type Settings interface {
+	DeltaOps(key string, delta settings.ItemChanges) ([]txn.Op, error)
+}
+
+type spaceRenameModelOp struct {
+	st       RenameSpaceState
+	space    RenameSpace
+	settings Settings
+	toName   string
+}
+
+func (o *spaceRenameModelOp) Done(err error) error {
+	return err
+}
+
+func NewRenameSpaceModelOp(settings Settings, st RenameSpaceState, space RenameSpace, toName string) *spaceRenameModelOp {
+	return &spaceRenameModelOp{
+		st:       st,
+		settings: settings,
+		space:    space,
+		toName:   toName,
+	}
+}
+
+type renameSpaceStateShim struct {
```

Comment:
Having this here breaks up the `spaceRenameModelOp` definition - part it is above, part below.